attest/attest-tool/attest-tool.go

```diff
@@ -122,8 +152,88 @@ func runCommand(tpm *attest.TPM) error {
 		}
 		fmt.Printf("%x\n", b)
 
+	case "dump":
+		dumpData, err := runDump(tpm)
+		if err != nil {
+			return fmt.Errorf("runDump() failed: %v", err)
+		}
+		return json.NewEncoder(os.Stdout).Encode(dumpData)
+
 	default:
 		return fmt.Errorf("no such command %q", flag.Arg(0))
 	}
 	return nil
 }
+
+func runDump(tpm *attest.TPM) (*dump, error) {
+	var (
+		out dump
+		err error
+	)
+
+	out.Static.TPMVersion = tpm.Version()
+	if out.Static.EKPem, err = rsaEKPEM(tpm); err != nil {
+		return nil, err
+	}
+
+	k, err := tpm.MintAIK(nil)
+	if err != nil {
+		return nil, fmt.Errorf("MintAIK() failed: %v", err)
+	}
+	defer k.Close(tpm)
+	out.AIK = k.AttestationParameters()
+
+	// Get a quote.
+	if out.Quote.Nonce, err = hex.DecodeString(*nonceHex); err != nil {
+		return nil, err
+	}
+	out.Quote.Alg = attest.HashSHA1
+	if *useSHA256 {
+		out.Quote.Alg = attest.HashSHA256
+	}
+	q, err := k.Quote(tpm, out.Quote.Nonce, out.Quote.Alg)
+	if err != nil {
+		return nil, fmt.Errorf("Quote() failed: %v", err)
+	}
+	out.Quote.Quote = q.Quote
+	out.Quote.Signature = q.Signature
+
+	// Get log information.
+	if out.Log.Raw, err = tpm.MeasurementLog(); err != nil {
+		return nil, err
+	}
+	if out.Log.PCRs, out.Log.PCRAlg, err = tpm.PCRs(); err != nil {
+		return nil, err
+	}
+	return &out, nil
+}
+
+func rsaEKPEM(tpm *attest.TPM) ([]byte, error) {
+	eks, err := tpm.EKs()
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		pk  *rsa.PublicKey
+		buf bytes.Buffer
+	)
+	for _, ek := range eks {
+		if ek.Cert != nil && ek.Cert.PublicKeyAlgorithm == x509.RSA {
+			pk = ek.Cert.PublicKey.(*rsa.PublicKey)
+			break
+		} else if ek.Public != nil {
+			pk = ek.Public.(*rsa.PublicKey)
+			break
+		}
+	}
+
+	if pk == nil {
+		return nil, errors.New("no EK available")
+	}
+
+	if err := pem.Encode(&buf, &pem.Block{Type: "RSA PUBLIC KEY", Bytes: x509.MarshalPKCS1PublicKey(pk)}); err != nil {
+		return nil, fmt.Errorf("pem.Encode() failed: %v", err)
```

Comment:
nit: seen this a few times, but generally avoid referencing function names in error messages. just describe what failed

```
return nil, fmt.Errorf("pem encoding RSA key failed: %v")
```